src/app/InteractionModelEngine.h

```diff
@@ -48,7 +48,7 @@
 #include <app/util/basic-types.h>
 
 #define CHIP_MAX_NUM_COMMAND_HANDLER 1
-#define CHIP_MAX_NUM_COMMAND_SENDER 1
+#define CHIP_MAX_NUM_COMMAND_SENDER 4
```

Comment:
Some thoughts:

1. This needs to be configurable per-device-build.  How many senders are needed depends on the device functionality.  Devices that don't implement any client sides of clusters don't need these things at all, right?  A device that actually includes CHIPDeviceController might want some number, and that number depends on how parallel it expects its access to be.
2. The model of "one sender per CHIPDevice" is not right.  Most CHIPDevices are not sending commands most of the time.  

In practice, "4" is not enough for our internal (manual) smoketests to pass.  But those are running on iOS, not a constrained device, which is where "configurable per-device-build" comes in....

Reply:
To add to this, we need to make sure that we aren't thinking about this only from the constrained device perspective.  We have a single stack that is meant to support both constrained and unconstrained devices.

CHIP solves the port exhaustion problem and can manage all session traffic through a single port.  The node ID space is 64-bits.  The practical limits to the number of devices a controller can concurrently support are really just: (1) memory, (2) network capacity and (3) the scalability of mDNS.  The spec doesn't preclude very large nodes from communicating with a very large number of peers.  The sdk should support such use cases.

Reply:
I think it would be good to explore a model where we don't always have to allocate a Sender out of the pool inside the IM Engine, but rather, permit clients to make a new Sender on their own allocated on heap dynamically as they need to on platforms that permit that (which would result in them sizing that pool to 0).

Arguably, if you look at InteractionModelEngine::NewCommandSender, all it's doing is Initing it with mpExchangeMgr (which is known to the application) and mpDelegate (which can be retrieved from InteractionModelEngine::GetInstance()->mpDelegate, assuming friend class).

This avoid this whole sizing problem on those platforms, and permits a more scalable pattern as fits their needs.

Reply:
That raises the question: why bother with a pool inside the IM engine at all?  Clients that want a pool can do their own pool...